File: binsV1/binV9/seventh_problem.js
import fs from "fs";

const staffs = JSON.parse(fs.readFileSync("input_data/staff.json", "utf8"));

const latestFeedbacks = staffs.map((staff) => {
  const feedbacks = staff?.result?.feedback;

  const latestFeedback = feedbacks?.reduce((latest, current) => {
    return new Date(latest.date) > new Date(current.date) ? latest : current;
  }, feedbacks[0]);

  return {
    id: staff.id,
    full_name: staff.full_name,
    email: staff.email,
    feedback: latestFeedback?.comment ?? "No available feedback yet.",
  };
});

console.log(JSON.stringify({ data: latestFeedbacks }, null, 2));

File: bins/binsV1/binV8/seventh_problem.rb
require 'json'
require 'date'

staffs = JSON.parse(File.read('input_data/staff.json'), symbolize_names: true)

latest_feedbacks = staffs.map do |staff|
  feedbacks = staff.dig(:result, :feedback).max_by do |feedback|
    Date.parse(feedback[:date])
  end

  {
    id: staff[:id],
    full_name: staff[:full_name],
    email: staff[:email],
    latest_feedbacks: feedbacks&.fetch(:comment, 'No available feedback yet.')
  }
end

puts JSON.pretty_generate({ data: latest_feedbacks })
